### PR TITLE
Added language option for RunAction

### PR DESCRIPTION
**Pull request by zzzkk**
### Short description 📝

Added optional `language` option for `RunAction` (consistent with convention in `TestAction`)

Screenshot 2021-07-26 at 16 34 27

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Tests/TuistGeneratorTests/Generator/SchemeDescriptorsGeneratorTests.swift
Comment (fortmarek) [inline Tests/TuistGeneratorTests/Generator/SchemeDescriptorsGeneratorTests.swift:759]: nit: I'd also add `XCTAssertNil(result.language)` to at least one of the other test cases to assert the default is what we expect

**Review by kwridan — APPROVED** [review]
Thanks for this contribution @zzzkk It's a welcome addition to the functionality 👍 

Some further thoughts:

- The fixture https://github.com/tuist/tuist/blob/main/projects/tuist/fixtures/ios_app_with_custom_scheme/App/Project.swift#L15 could be extended to also use the language option, this can act as a nice show case of the functionality.

- Seeing we're trying to be consistent with `TestAction`, this next suggestion could be a consideration for #2981 


`SchemeLanguage` type could be introduced and used instead of `String` _(in `ProjectDescription` at least)_, this would allow us to pre-define some languages in Tuist for use especially for some of those special ones at the bottom of the list

languages


e.g.

```swift
public struct SchemeLanguage: Codable {
    var identifier: String
    public init(identifier: String) {
       self.identifier = identifier
    }
}


// To support using string literals directly when specifying languages
public extension SchemeLanguage: ExpressibleByStringLiteral {
  // ...
}

// Pre-defined languages
public extension SchemeLanguage {
    static var doubleLengthPseudoLanguage: SchemeLanguage {
        SchemeLanguage(identifier: "IDELaunchSchemeLanguageDoubleLocalizedStrings")
    }
    // ...
}
```

File: Sources/TuistLoader/Models+ManifestMappers/RunAction+ManifestMapper.swift
Comment (fortmarek) [inline Sources/TuistLoader/Models+ManifestMappers/RunAction+ManifestMapper.swift:28]: nit: This can be probably used directly, I don't think defining as its own variable brings any real value here

File: Sources/ProjectDescription/RunAction.swift
Comment (fortmarek) [inline Sources/ProjectDescription/RunAction.swift:20]:
```suggestion
    /// App Language.
```

**Review by fortmarek — APPROVED** [review]
Thanks a lot @zzzkk, this is a great addition 🙌

**Comment by fortmarek** [comment]
Thanks @kwridan for the additional review.

> Seeing we're trying to be consistent with TestAction, this next suggestion could be a consideration for 2.0 deprecations #2981

I wholeheartedly agree - it'd be great to offer things out-of-the-box in a type-safe manner, rather than forcing users to dig for what the real value should be. I have added it to the attached issue.

**Review by kwridan — APPROVED** [review]
Thank you for all the updates @zzzkk, it's looking great!

The `TestAction` change is backward compatible with any string literal use 👍  

```swift
// ProjectDescriptionHelpers

func makeTestAction() -> TestAction {
   TestAction(
         // ...
         language: "pl"
    )
}
```

However it could be a breaking change if anyone used string variables in their manifests

```swift
// ProjectDescriptionHelpers

func makeTestAction(language: String) -> TestAction {
   TestAction(
         // ...
         language: language
    )
}
```

We could offer two initialisers for `TestAction`, marking the `String?` one as deprecated for that possible use case in case anyone is using it. Alternatively could make a note in the change log as **breaking** or **potentially breaking** 😅 - @fortmarek thoughts?

**Comment by fortmarek** [comment]
> We could offer two initialisers for TestAction, marking the String? one as deprecated for that possible use case in case anyone is using it.

Yeah, I think this is the best way going forward. It's always better to make the change non-breaking if we can. We will remove the second init when doing the bump to 2.0.

I can also see that the format checks are failing, you can run `./fourier format all --fix` to fix those, @zzzkk 🙂

**Comment by fortmarek** [comment]
@all-contributors add @zzzkk for code

**Comment by pepibumur** [comment]
@zzzkk let me know if you need an extra hand to push this one to the finish line :)

**Comment by zzzkk** [comment]
> @zzzkk let me know if you need an extra hand to push this one to the finish line :)

Hi @pepibumur, 
I will do my best to finish it today, I'm little busy last days 😉 
thanks @kwridan @fortmarek  for reviews 🙌

**Comment by zzzkk** [comment]
> We could offer two initialisers for `TestAction`, marking the `String?` one as deprecated for that possible use case in case anyone is using it

@kwridan in fact we can't have two initializers with different `language` type (`String?` and `SchemeLanguage?`)
```
   TestAction(
         // ...
         language: nil
    )
```
will be ambiguous

I think there are two options now
1. Revert changes for TestAction and leave it like it is 
2. Make `SchemeLanguage` not optional but with optional `identifier` for example `language: SchemeLanguage = SchemeLanguage(identifier: nil)`

let me know what you think @kwridan @fortmarek

**Comment by kwridan** [comment]
> @kwridan in fact we can't have two initializers with different language type (String? and SchemeLanguage?)

Ah good point missed this!

Having had another look at `TestAction` I see we have one private initialiser, one public initialiser and a few public static functions which also act as initialisers. Perhaps we keep the public initialiser as is `String?`, and convert the private initialiser to `SchemeLanguage?` and add `language: SchemeLanguage?` to all the static functions? That should resolve the ambiguity for the older public initialiser while keeping the newer static function leverage the new types. Does that sound reasonable?

**Review by fortmarek — APPROVED** [review]
💜 This is good to go from my side, I'll let @kwridan to have one last look before merging. Thanks a lot, @zzzkk! I have also invited to the Tuist contributors team, so you don't have to create a fork next time you want to contribute 🙂

**Review by kwridan — COMMENTED** [review]
Thank you for the updates @zzzkk!!

My apologies I forgot that the language option is not applicable when using test plans as it's defined in the test plan itself and hidden from the Xcode scheme editor 🤦‍♂️

testplans

Given that `language` isn't applicable in the `testPlan` static methods we can remove that update and tackle the `TestAction` consolidation updates separately later on as part of 2.0 deprecations #2981 where we have some room for making breaking changes.

Apologies again for leading down this path and thank you for your patience with this PR!!

**Review by kwridan — APPROVED** [review]
Thanks @zzzkk !

**Pull request merged**
